crates/biome_js_analyze/src/analyzers/nursery/use_as_const_assertion.rs

```diff
@@ -0,0 +1,280 @@
+                let asserted_literal = expr.ty().ok()?;
+                let range = check_literal_match(literal, &asserted_literal)?;
+                Some(RuleState::AngleBracketAssertion(range))
+            }
+            Query::JsVariableDeclarator(decl) => {
+                let literal = decl.initializer()?.expression().ok()?;
+                let literal = literal.as_any_js_literal_expression()?;
+                let type_annotation = decl.variable_annotation()?;
+                let type_annotation = type_annotation.as_ts_type_annotation()?;
+                let type_annotation = type_annotation.ty().ok()?;
+                let range = check_literal_match(literal, &type_annotation)?;
+                Some(RuleState::TypeAnnotation(range))
+            }
+            Query::JsPropertyClassMember(member) => {
+                let literal = member.value()?.expression().ok()?;
+                let literal = literal.as_any_js_literal_expression()?;
+                let property_annotation = member
+                    .property_annotation()?
+                    .as_ts_type_annotation()?
+                    .ty()
+                    .ok()?;
+                let range = check_literal_match(literal, &property_annotation)?;
+                Some(RuleState::TypeAnnotation(range))
+            }
+        }
+    }
+
+    fn diagnostic(_: &RuleContext<Self>, state: &Self::State) -> Option<RuleDiagnostic> {
+        match state {
+            RuleState::AsAssertion(range) => Some(
+                RuleDiagnostic::new(
+                    rule_category!(),
+                    range,
+                    markup! {
+                        "You should use "<Emphasis>"as const"</Emphasis>" instead of "<Emphasis>"as"</Emphasis>" with a literal type."
+                    },
+                ).note(markup! {""<Emphasis>"as const"</Emphasis>" is simpler and doesn't require retyping the value."})
+            ),
+            RuleState::AngleBracketAssertion(range) => {
+                Some(RuleDiagnostic::new(
+                    rule_category!(),
+                    range,
+                    markup! {
+                        "You should use "<Emphasis>"as const"</Emphasis>" instead of angle bracket type assertion."
+                    }
+                ).note(markup! {"The angle bracket assertion can occasionally be confused with JSX syntax, so using the "<Emphasis>"as const"</Emphasis>" is a more clear and preferable alternative."}))
+            },
+            RuleState::TypeAnnotation(range) => Some(
+                RuleDiagnostic::new(
+                    rule_category!(),
+                    range,
+                    markup! {
+                        "You should use "<Emphasis>"as const"</Emphasis>" instead of type annotation."
+                    }
+                ).note(markup! {""<Emphasis>"as const"</Emphasis>" is simpler and doesn't require retyping the value."})
```

Comment:
This could be nice to factorize this part. You could do that by converting the `if` statement to an expression in which the result is a tuple `(message, note)`.